### PR TITLE
feat: ability to pass base64 encoded signature keys

### PR DESCRIPTION
**Pull request by Aashiq-J**
### Description

Add a new variable to accept base64 encoded value of the signature keys which will make it compatible with the schematics.

#### Changes that affect the core Terraform module or submodules
- [ ] Bug fix
- [x] New feature
- [ ] Dependency update

#### Changes that don't affect the core Terraform module or submodules
- [ ] Examples or tests (addition or updates of examples or tests)
- [ ] Documentation update
- [ ] CI-related update (pipeline, etc.)
- [ ] Other

#### Release required?
Identify the type of release. For information about the changes in a semantic versioning release, see [Release versioning](https://terraform-ibm-modules.github.io/documentation/#/versioning).

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [x] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

##### Release notes content

If a release is required, replace this text with information that users need to know about the release. Write the release notes to help users understand the changes, and include information about how to update from the previous version.

Your notes help the merger write the commit message for the PR that is published in the release notes for the module.

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### Merge actions for mergers

- Use a relevant [conventional commit](https://www.conventionalcommits.org/) message that is based on the PR contents and any release notes provided by the PR author. The commit message determines whether a new version of the module is needed, and if so, which semver increment to use (major, minor, or patch).
- Merge by using "Squash and merge".

### CODE REVIEW

File: variables.tf [hunk 1]
```diff
@@ -91,6 +91,17 @@ variable "admins" {
...
 
+variable "base64_encoded_admins" {
+  type = list(object({
+    name  = string # max length: 30 chars
+    key   = string #  base64 encoded value of signature key files if key files are created using TKE CLI and are not using a third-party signing service
+    token = string # sensitive: the administrator password/token to authorize and access the corresponding signature key file
+  }))
+  default     = []
+  sensitive   = true
+  description = "A list of administrators for the instance crypto units. See [instructions](https://github.com/terraform-ibm-modules/terraform-ibm-hpcs#before-you-begin) to create administrator signature keys. You can set up to 8 administrators. Required if auto_initialization_using_recovery_crypto_units set to true. The signature keys should passed as base64 encoded values."
```
Comment (SirSpidey) [inline variables.tf]: I think we should assume that descriptions are plain text, not markdown. I think that not all places handle MD correctly.

I shortened a bit and moved the readme stuff to the end, but I think more context is needed. How is this different from the `admins` variable. Are they mutually exclusive? Do users need to know which to use?  

```suggestion
  description = "A list of up to 8 administrators for the instance crypto units. Required if auto_initialization_using_recovery_crypto_units is set to true. Pass the signature keys as base64 encoded values. For information about administrator signature keys, see the readme file."
```

File: module-metadata.json [hunk 2]
```diff
@@ -26,6 +26,17 @@
         "line": 34
       }
     },
+    "base64_encoded_admins": {
+      "name": "base64_encoded_admins",
+      "type": "list(object({\n    name  = string # max length: 30 chars\n    key   = string #  base64 encoded value of signature key files if key files are created using TKE CLI and are not using a third-party signing service\n    token = string # sensitive: the administrator password/token to authorize and access the corresponding signature key file\n  }))",
+      "description": "A list of administrators for the instance crypto units. See [instructions](https://github.com/terraform-ibm-modules/terraform-ibm-hpcs#before-you-begin) to create administrator signature keys. You can set up to 8 administrators. Required if auto_initialization_using_recovery_crypto_units set to true. The signature keys should passed as base64 encoded values.",
```
Comment (SirSpidey) [inline module-metadata.json]: see comment in variables.tf

**Review by SirSpidey — COMMENTED** [review]
I suggest that the new variable needs more context to distinguish from `admins` and made other suggestions. 

The following are suggestions for other variables. If you don't want to make the changes in this PR, I understand. 

1. Every description that has *The number of xx that is required...*, change to `The number of xx required...` (remove "that is").
2. For `service_endpoints`, suggest this update:

    `The service_endpoints to access your service instance. Used only if auto_initialization_using_recovery_crypto_units is set to true. Can be set to private-only if Terraform has access to the private endpoints. Default value is public-and-private.`
3. `signature_server_url` description: Update last sentence as above
    `Used only if auto_initialization_using_recovery_crypto_units is set to true.`

File: variables.tf (hunk 1 above)
Comment (Aashiq-J) [reply variables.tf]: Yea, they are mutually exclusive. Only one can be passed at a time.

**Review by SirSpidey — COMMENTED** [review]
Thanks for the changes. 
I'm assuming that it's not obvious to everyone which variable to use. So I think we need to provide guidance. If you and others don't agree, I can live with what we have now though.
 
I think each admins variable should point to the other. I'm not sure what the conditions are for choosing base64. 

The statements could be something like this, 

for `admins`:
> If you deploy with IBM Cloud Schematics (change to what the reason is) use the base64_encoded_admins input instead of this one.

for base64_encoded_admins:
> If you don't need base64 encoding, use the admins input.

**Comment by Aashiq-J** [comment]
@SirSpidey ,
Thanks for the suggestion. I'm not sure if pointing to other variable is required.
The idea of having two variable is because, the hpcs module needs a local signature file present in the machine filesystem while the terraform deploy and initialises the hpcs service.
So, if we run the terraform from a machine which has those signature key file we can use `admins` and pass the absolute path of the signature files.

But, when using schematics to deploy hpcs, there is no option to upload the signature files to the schematics workspace, so we added an alternative to base64 encode the file and pass that encoded value as part of the `base64_encoded_admins` variable. So, its not mandatory that `base64_admins` can only be used when using schematics workspace. But, to deploy hpcs using  schematics workspace `base64_admins` is required.

**Comment by SirSpidey** [comment]
@Aashiq-J It's not obvious to me how everyone using the module would understand when they have to use `admins` vs `base64_encoded_admins`. 

I'll approve, but I'm approving only the content. I think you should also get a developer to review the code.

**Review by SirSpidey — DISMISSED** [review]
Approving based on Aashiq's conclusion that we don't need to add context to the variables to explain when to use each.

**Review by ocofaigh — APPROVED** [review]

**Pull request merged**